Create bowling.rb

ngmysh opened the pull request:
ボウリングのスコア計算プログラムを作成

ngmysh committed:
Create bowling.rb

ボウリングのスコア計算プログラムを作成

ngmysh committed:
Change style

if文の戻り値を使って値を代入するときのスタイルを変更

maedana commented on 04.bowling/bowling.rb:
```diff
@@ -0,0 +1,48 @@
+    else
+      score.to_i
+    end
+end
```
`map`使って初期化なしで代入までやるとスッキリかけますね。
```ruby
shots = scores.map do |score|
  if score == 'X'
    10
  else
    score.to_i
  end
end
```
ここまで来ればブロック内部の単純なif/elseは三項演算子を使いつつ一行で書いてもいいかもです。
```ruby
shots = scores.map { |score| score == 'X' ? 10 : score.to_i }
```

ngmysh replied:
この課題の前にチェリー本を読んでいたのにmapで配列を初期化するテクニックをすっかり忘れていました。。。
修正しておきます！

maedana commented on 04.bowling/bowling.rb:
```diff
@@ -0,0 +1,48 @@
+  # トータルのフレームに現在のフレームを追加する条件は次の通り
+  # ・(最終フレームではない) && (ストライク || フレームの2投目)
+  # ・最後の1投
+  next unless ((frames.size != 9) && (shot == 10 || tmp_frame.size == 2)) || (index == shots.size - 1)
```
> 少し気になったところはrubocopに怒られてnext unlessを使った箇所です。
> rubyに慣れている人からすると、この箇所の条件分岐はifよりもnext unlessの方が見やすいのでしょうか？

と言及されてましたが、確かにこのままでは読みにくいですね。現在のロジックを活かすなら例えば以下のように分解するとどうでしょう?
```ruby
  next if frames.size != 9
  next if shot != 10 && tmp_frame.size != 2
  next if index != shots.size - 1
```

さらにそれぞれをメソッド呼び出しにするとリーダブルになるかもですね。実装まで落とし込んで試してませんが、例えば
```ruby
  next if last_frame?(frames)
  next if strike?(shot) && second_throw_in_frame?(tmp_frame)
  next if last_throw?(shots, index)
```
のような感じでしょうか。あとはそもそももっとシンプルなロジックを考えてみる、とかですかね。(現時点で無理にロジック変更しなくていいです。合格したあとで他の方のコードを見てみるなどするといいかなと)

ngmysh replied:
複数行に分けてわかりやすくするロジックを考えていて頭がこんがらがってきてしまったので、この部分はロジックをシンプルに変更することにしました。
ご指摘いただいたように、複数の関数を追加してリーダブルになったと思ったのでコメントは削除することにしました。

maedana commented on 04.bowling/bowling.rb:
```diff
@@ -0,0 +1,48 @@
+          frames[index + 1][0] + frames[index + 2][0]
+        else
+          frames[index + 1][0] + frames[index + 1][1]
+        end
```
以下のようにまとめてもいいかなと思いました。
```ruby
frames[index + 1][0] + (frames[index + 1][1] || frames[index + 2][0])
```

ngmysh replied:
めちゃくちゃシンプルになって気持ちいです。
この部分も修正しておきます！

ngmysh committed:
Clean some code to simple and readable and Add test code

配列の初期化にmapを使用するように変更、複雑な条件を使用していたロジックを複数行の条件分岐に変更、単純な条件分岐を一行にまとめる実装に変更
テストコードを追加
